tsuru/resource_tsuru_app_env.go

```diff
@@ -0,0 +1,406 @@
+		if len(envs) > 0 {
+			varNames := getVarNames(envs)
+			resp, err := client.AppApi.EnvUnset(ctx, app, varNames, true)
+			if err != nil {
+				return resource.NonRetryableError(err)
+			}
+
+			defer resp.Body.Close()
+			body, err := ioutil.ReadAll(resp.Body)
+			if err != nil {
+				return resource.NonRetryableError(err)
+			}
+
+			if resp.StatusCode != http.StatusOK {
+				if isLocked(string(body)) {
+					return resource.RetryableError(errors.Errorf("App locked"))
+				}
+				return resource.NonRetryableError(errors.Errorf("unable to add cname, error code: %d", resp.StatusCode))
+			}
+		}
+		return nil
+	})
+}
+
+func getRemoteEnvironment(ctx context.Context, client *tsuru_client.APIClient, app string) (private, public []tsuru_client.Env, err error) {
+	envs, _, err := client.AppApi.EnvGet(ctx, app, nil)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	for _, env := range envs {
+		if isReservedEnv(env.Name) {
+			continue
+		}
+		if env.Public {
+			public = append(public, env)
+		} else {
+			private = append(private, env)
+		}
+	}
+
+	return
+}
+
+func diffEnvSetData(newEnvs, oldEnvs []tsuru_client.Env) (update, remove []tsuru_client.Env) {
+	update = []tsuru_client.Env{}
+	remove = []tsuru_client.Env{}
+
+	for _, newEnv := range newEnvs {
+		if curEnv, ok := envInSet(newEnv.Name, oldEnvs); ok {
+			if curEnv.Value != newEnv.Value {
+				update = append(update, newEnv)
+			}
+		} else {
+			update = append(update, newEnv)
+		}
+	}
+
+	for _, oldEnv := range oldEnvs {
+		if _, ok := envInSet(oldEnv.Name, newEnvs); !ok {
+			remove = append(remove, oldEnv)
+		}
+	}
+
+	return
+}
+
+func envInSet(envName string, envs []tsuru_client.Env) (*tsuru_client.Env, bool) {
+	for _, e := range envs {
+		if e.Name == envName {
+			return &e, true
+		}
+	}
+	return nil, false
+}
+
+func getVarNames(envs []tsuru_client.Env) []string {
+	names := []string{}
+	for _, e := range envs {
+		names = append(names, e.Name)
+	}
+	return names
+}
+
+func envsFromResource(envvars interface{}) (*tsuru_client.EnvSetData, *tsuru_client.EnvSetData) {
+	evs := envvars.(*schema.Set)
+
+	private := &tsuru_client.EnvSetData{
+		Envs:    []tsuru_client.Env{},
+		Private: true,
+	}
+
+	public := &tsuru_client.EnvSetData{
+		Envs: []tsuru_client.Env{},
+	}
+
+	for _, raw := range evs.List() {
+		e := raw.(map[string]interface{})
+		env := tsuru_client.Env{
+			Name: e["name"].(string),
+		}
+		if v, ok := e["value"]; ok && v != "" {
+			env.Value = v.(string)
+			public.Envs = append(public.Envs, env)
+			continue
+		}
+		if v, ok := e["sensitive_value"]; ok && v != "" {
+			env.Value = v.(string)
+			private.Envs = append(private.Envs, env)
+			continue
+		}
+	}
+
+	return private, public
+}
+
+func isReservedEnv(variable string) bool {
+	reserved := map[string]bool{
+		"TSURU_APPDIR":    true,
+		"TSURU_APPNAME":   true,
+		"TSURU_APP_TOKEN": true,
+		"TSURU_SERVICES":  true,
+	}
+	_, ok := reserved[variable]
```

Comment:
@pedrokiefer we could use all strings starting with `TSURU_` as a reserved environment name